Distionary post (#1222)
ropensci/roweb3: content/blog/2026-01-20-introducing-distionary/index.Rmd

---
title: Introducing distionary for Building and Probing Distributions
author:
  - Vincenzo Coia
date: '2026-01-20'
slug: introducing-distionary
tags:
  - tech notes
  - distributions
  - peer-reviewed
  - community
  - distionary
  - probaverse
  - packages
package_version: 0.1.0
social: A post about a new probability distribution framework in R with distionary and the wider probaverse.

> description: The distionary package is on CRAN and brings a new way to build and probe probability distributions in R, forming the building blocks of the probaverse ecosystem.

Reviewer: Because you'll be cross posting on the R Consortium Blog, we'll have you add a preface here noting that. We can put in the link once you know what it will be.

Then for the blog post on the R Consortium, we'll ask that you have a similar note `Cross-posted on the [**rOpenSci Blog**](https://ropensci.org/blog/2026/01/20/introducing-distionary/).`

This link doesn't work yet, but will be the official link as long as the date doesn't change 😁 

```suggestion
description: The distionary package is on CRAN and brings a new way to build and probe probability distributions in R, forming the building blocks of the probaverse ecosystem.
preface: "Cross-posted on the [**R Consortium Blog**](LINK)"
```

> editor: ~

Reviewer: ```suggestion
editor: Steffi LaZerte
params:
  doi: "10.59350/pg1t4-ty826"
```

# And to disable symbols before output
knitr::opts_chunk$set(fig.path = "", comment = "", fig.align = "center")

# knitr hook to make images output use Hugo options
knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "`{{< figure src=",
      '"', x, '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}`{=html}\n"
    )
  }
)

# knitr hook to use Hugo highlighting options
knitr::knit_hooks$set(
  source = function(x, options) {
  hlopts <- options$hlopts
    paste0(
      "```r ",
      if (!is.null(hlopts)) {
      paste0("{",
        glue::glue_collapse(
          glue::glue('{names(hlopts)}={hlopts}'),
          sep = ","
        ), "}"
        )
      },
      "\n", glue::glue_collapse(x, sep = "\n"), "\n```\n"
    )
  }
)
```

The [distionary](https://docs.ropensci.org/distionary/) package is newly available on CRAN.
It allows you to make probability distributions quickly -- either from a few inputs or from its built-in library -- and then probe them in detail.

> These distributions form the building blocks that piece together advanced statistical models with the wider [probaverse](https://probaverse.com) ecosystem, which is built to release modelers from low-level coding so production pipelines stay human-friendly.

Reviewer: The probaverse sounds pretty cool! Consider adding a sentence or two to explain a bit more about what it does? Perhaps reference a couple of the other packages, like the ones you use in this post (i.e. famish).

> These distributions form the building blocks that piece together advanced statistical models with the wider [probaverse](https://probaverse.com) ecosystem, which is built to release modelers from low-level coding so production pipelines stay human-friendly.

Reviewer: ```suggestion
These distributions form the building blocks that piece together advanced statistical models with the wider [probaverse](https://probaverse.com) ecosystem, which is designed to release modelers from low-level coding so production pipelines stay human-friendly.
```

  "bottomright",
  legend = c("Empirical", "Fitted Gamma"),
  col = c("black", "red"),
  lty = 1
)
```

These textbook distributions become much more useful once they become building blocks for building up a system.
For example, they could form predictive distributions in a machine learning context, or be related to other variables.
This is what the probaverse seeks to make possible.

## Feature 2: friendly towards tidy tabular workflows

First, load the tidyverse to activate tidy tabular workflows.
And yes, probaverse is named after the tidyverse because it aims to be a "tidyverse for probability".

```{r, message = FALSE, warning = FALSE}
library(tidyverse)
```

You can safely ignore this next chunk unless you want to see how I'm wrangling some financial data for you.

```{r}
# Wrangle the stocks data frame using tidyverse.
stocks <- as_tibble(EuStockMarkets) |> 
  mutate(across(everything(), \(x) 100 * (1 - x / lag(x)))) |> 
  drop_na()
```

The `stocks` data I've wrangled is a table of _daily percent loss_ for four major European stock indices.
The dates don't matter for this example, so they've been omitted.

```{r}
stocks
```

First, let's focus on the DAX stock index.

> Fit an empirical distribution like last time (notice I'm using a data mask in `dst_empirical()` this time).

Reviewer: Consider explaining data masking in a footnote. Many readers will likely understand what it is, but not know the term. Maybe something like the following?

```suggestion
Fit an empirical distribution like last time (notice I'm using a data mask[^1] in `dst_empirical()` this time).

[^1]: Meaning I'm referring directly to the column 'DAX' without `stocks$` as in our above examples.
```

```{r empirical_dax, hugoopts=list(alt = "Empirical CDF of DAX stock index daily losses.")}
# Fit an empirical distribution to the DAX stock index.
dax <- dst_empirical(DAX, data = stocks)
# Inspect the CDF.
plot(dax, xlab = "Daily Loss (%)")
```

> You can easily calculate some standard quantiles in tabular format so that the inputs are placed alongside the calculated outputs: just use the `enframe_` prefix instead of `eval_`.

Reviewer: Consider reminding the reader where `eval_` came from (I used Ctrl-F to remind myself :D)
```suggestion
You can easily calculate some standard quantiles in tabular format so that the inputs are placed alongside the calculated outputs: just use the `enframe_` prefix instead of `eval_` as we did above with the Null distribution.
```

```{r}
enframe_quantile(dax, at = c(0.25, 0.5, 0.75), arg_name = "prob")
```

Or, more to the point here -- and appealing to probaverse's soft spot for risk-focused work -- you can calculate return levels (also known as "Value at Risk" in financial applications) for specific return periods.

> If you don't know what these are, they are just fancy names for quantiles.

Reviewer: Love it!

)
```

The tabular output becomes even more powerful when inserted into a table of models, because it facilitates comparisons and trends.
To demonstrate, build a model for each stock. First, lengthen the data for this task.

```{r}
# Lengthen the data using tidyverse.
stocks2 <- stocks |> 
  pivot_longer(everything(), names_to = "stock", values_to = "daily_loss_pct")
# Inspect
stocks2
```

Build a model for each stock using a `group_by` + `summarise` workflow from the tidyverse (please excuse the current need to wrap the distribution in `list()`). Notice that distributions become table entries, indicated here by `<dst>`.

```{r}
# Create an Empirical distribution for each stock.
models <- stocks2 |> 
  group_by(stock) |> 
  summarise(model = list(dst_empirical(daily_loss_pct)))
# Inspect
models
```

Now you can use a tidyverse workflow to calculate tables of quantiles for each model, and expand them.
In fact, this workflow is common enough that I'm considering a dedicated verb for it.

```{r}
return_levels <- models |>
  mutate(df = map(
    model, enframe_return,
    at = return_periods,
    arg_name = "return_period",
    fn_prefix = "daily_loss_pct"
  )) |> 

>   unnest(df) |> 

Reviewer: If you like, to match the above styles (I have [air](https://tidyverse.org/blog/2025/02/air/) operating so it keeps trying to shift your code when I work locally 😁)
```suggestion
  mutate(
    df = map(
      model,
      enframe_return,
      at = return_periods,
      arg_name = "return_period",
      fn_prefix = "daily_loss_pct"
    )
  ) |>
  unnest(df) |>
```

  select(!model)
# Inspect
print(return_levels, n = Inf)
```

The result is a tidy dataset that's ready for most analyses.
For example, you can easily plot a comparison of the return levels of each stock.
I make these plots all the time to facilitate risk-informed decision-making.

```{r return_period_plot, hugoopts=list(alt="Return Level Plot for Daily Loss Percentages of Stock Indices.")}
return_levels |> 
  mutate(stock = fct_reorder2(stock, return_period, daily_loss_pct)) |>
  ggplot(aes(return_period, daily_loss_pct, colour = stock)) +
  geom_point() +
  geom_line() +
  theme_bw() +
  scale_x_log10(
    "Return Period (days)",
    minor_breaks = c(1:10, 1:10 * 10, 1:10 * 100)
  ) +
  scale_y_continuous("Daily Loss", label = scales::label_number(suffix = "%")) +
  annotation_logticks(side = "b") +
  scale_colour_discrete("Stock Index")
```

## Feature 3: make the distribution you need

> You can make your own distributions with distionary without needing to specify every property in order to retrieve these properties.

Reviewer: I found this wording a bit circular, but I'm not sure my suggestion is any better. What do you think?

```suggestion
You can make your own distributions with distionary without needing to specify every property in order to create a distribution from which you can retrieve other properties.
```

Let's say you need an Inverse Gamma distribution but it's not available in distionary.

> Currently, distionary assumes you'll at least provide the density and CDF; you could retrieve these from the extraDistr package (functions `dinvgamma()` and `pinvgamma()`).

Reviewer: CDF? Perhaps explain or link to https://en.wikipedia.org/wiki/Cumulative_distribution_function?
Also, add a link for the extraDistr package so readers can easily find that too!

> Plug them into the `distribution()` function and enjoy access to a variety of properties you didn't specify, like the mean, variance, skewness, and hazard function.

Reviewer: ```suggestion
Plug them into distionary's `distribution()` function and enjoy access to a variety of properties you didn't specify, like the mean, variance, skewness, and hazard function.
```

Notice I also check that the parameters are positive (cheers to the checkmate package).

```{r}
dst_invgamma <- function(alpha, beta) {
  checkmate::assert_number(alpha, lower = 0)
  checkmate::assert_number(beta, lower = 0)
  distribution(
    density = \(x) extraDistr::dinvgamma(x, alpha = alpha, beta = beta),
    cdf = \(x) extraDistr::pinvgamma(x, alpha = alpha, beta = beta),
    quantile = \(p) extraDistr::qinvgamma(p, alpha = alpha, beta = beta),
    random = \(n) extraDistr::rinvgamma(n, alpha = alpha, beta = beta),
    .name = "Inverse Gamma",
    .parameters = list(alpha = alpha, beta = beta),
    .vtype = "continuous",
  )
}
```

Now we can make that same Inverse Gamma distribution as before:

```{r}
ig2 <- dst_invgamma(5, 20)
# Inspect
ig2
```

By the way, this feature -- being able to inspect other distribution properties even when they are not specified -- is great for learning about probability.
That's because you can see the many ways distributions can be represented, not just by the usual density or probability mass functions seen in textbooks.

This feature also allows for extensibility of the probaverse.

> For example, the [distplyr](https://distplyr.probaverse.com/) package from the probaverse creates mixture distributions, for which a closed form for the quantile function does not exist.

Reviewer: What is a 'closed form'?
